nameresolution/consul/consul.go

```diff
@@ -81,96 +79,84 @@ type resolverConfig struct {
 
 // NewResolver creates Consul name resolver.
 func NewResolver(logger logger.Logger) nr.Resolver {
-	return newResolver(logger, resolverConfig{}, &client{})
+	return newResolver(logger, &client{})
 }
 
-func newResolver(logger logger.Logger, resolverConfig resolverConfig, client clientInterface) nr.Resolver {
+func newResolver(logger logger.Logger, client clientInterface) *resolver {
 	return &resolver{
 		logger: logger,
-		config: resolverConfig,
 		client: client,
 	}
 }
 
 // Init will configure component. It will also register service or validate client connection based on config.
-func (r *resolver) Init(metadata nr.Metadata) error {
-	var err error
-
+func (r *resolver) Init(metadata nr.Metadata) (err error) {
 	r.config, err = getConfig(metadata)
 	if err != nil {
 		return err
 	}
 
-	if err = r.client.InitClient(r.config.Client); err != nil {
+	err = r.client.InitClient(r.config.Client)
+	if err != nil {
 		return fmt.Errorf("failed to init consul client: %w", err)
 	}
 
-	// register service to consul
+	// Register service to consul
 	if r.config.Registration != nil {
-		if err := r.client.Agent().ServiceRegister(r.config.Registration); err != nil {
+		agent := r.client.Agent()
+
+		err = agent.ServiceRegister(r.config.Registration)
```

Comment:
I saw your comment about using `deregistercriticalserviceafter` - but that fix not working.

Should we rather deregister before registering?
Documentation for `ServiceDeregister(serviceID string) error` (https://developer.hashicorp.com/consul/commands/services/deregister) also recommends that it ` is meant to be paired with services register`.

It may not completely solve the issue but would certainly be a good practice most probably and may also help in solving issues like https://github.com/dapr/components-contrib/issues/2489

Reply:
That was actually what I first tried doing, and it didn't work because a service can have multiple instances if the app is scaled horizontally. Calling "ServiceDeregister" removed the entire service and so un-registered other instances of the app too.

I spent 2 hours on this yesterday and doesn't really look like Consul has any way to remove only one instance :(

Reply:
Ohh ok, got it.